docs: add getting started (alternative to #5519) (#5592)
gnolang/gno: docs/builders/getting-started.md

# Getting started with Gno.land

Gno.land is a Layer 1 blockchain where smart contracts are written in
**Gno**, a deterministic variant of Go. If you can write Go, you can
write Gno.

> Realms (`r/`) hold on-chain state, packages (`p/`) provide stateless

Author: Should be removed

libraries, and the GnoVM interprets everything. See
[What is Gno.land?](./what-is-gnolang.md) for the full picture.

This page walks you from zero to a working local chain and your first
on-chain transaction. Just want the commands? See [Quick Start](./quickstart.md).

:::tip
Try the **[Playground](https://play.gno.land)** to write Gno in your browser.
:::

## Install

The toolchain has three binaries:

| Binary    | What it is                                              |
|-----------|---------------------------------------------------------|
| `gno`     | the Gno language toolchain (format, test, run, mod init)|
| `gnokey`  | key management for interacting with the network         |
| `gnodev`  | local development environment with hot reload + web UI  |

```sh
curl -fsSL https://raw.githubusercontent.com/gnolang/gno/master/misc/install.sh | sh
```

Binaries land in `$HOME/.gno/bin`. The script is bash-only; Windows users should use
WSL or see [Other methods](./install.md) for source builds and Docker.

:::tip
`gno: command not found`? Add `$HOME/.gno/bin` to your `$PATH`:
`export PATH="$HOME/.gno/bin:$PATH"`.
:::

## Run locally with gnodev

> This section creates a realm, runs a local chain with `gnodev`, and

Author: Add better introduction

opens it in a browser in `gnoweb`.

### Declare the module path

```sh
gno mod init gno.land/r/myname/myrealm
```

This writes a `gnomod.toml` in the current directory, declaring the

> package's on-chain path. Use `gno.land/r/…` for realms (stateful) or

Author: Duplicate of first section

`gno.land/p/…` for packages (stateless — note that pure packages cannot
import realms; see [import rules](../resources/gno-packages.md#import-rules)).

### Write Gno code

> Add `.gno` files next to the freshly created `gnomod.toml`. We'll build

Author: Add linking word
Then, add .gno ...

increment it:

```gno
package myrealm

import "strconv"

var count int

func Increment(_ realm, n int) int {
	count += n
	return count
}

func Render(path string) string {
	return "Count: " + strconv.Itoa(count)
}
```

- **`var count int`** — top-level variables are automatically persisted
  to chain state after each transaction.
- **`Increment`** — the `_ realm` parameter makes the function
  ["crossing"](../resources/gno-interrealm.md), which is required for
  any function that modifies realm state. Callers pass `cross` as the
  first argument.
- **`Render`** — gnoweb calls this to display your realm in the
  browser. The signature must be `func Render(path string) string`.

Add a test file alongside it, `myrealm_test.gno`:

```gno
package myrealm

import "testing"

func TestIncrement(t *testing.T) {
	val := Increment(cross, 5)
	if val != 5 {
		t.Fatalf("expected 5, got %d", val)
	}
}
```

For deeper test patterns see the [Testing guide](../resources/gno-testing.md); a
fuller version of this counter lives at
[`examples/gno.land/r/demo/counter`](https://github.com/gnolang/gno/tree/master/examples/gno.land/r/demo/counter).

### Format, lint, and test

> 

Author: Explain what is it

```sh
gno fmt ./...     # rewrite .gno files in canonical style
gno lint ./...    # static checks for common mistakes
gno test ./...    # run _test.gno files
```

### Run a local chain

> 

Author: Missing introduction

> # Getting started with Gno.land

Reviewer: Add at the top of tutorialSidebar (before "Use Gno.land"). The two new pages are sidebar orphans on the rendered docs site. #5519 added the sidebar entry: this PR regresses on that.

Author: Fix: https://github.com/gnolang/gno/pull/5592/commits/160cf3bf8cd392271b5bd159bff648a4ac5f6e52

Gno.land is a Layer 1 blockchain where smart contracts are written in
**Gno**, a deterministic variant of Go. If you can write Go, you can
write Gno. See [What is Gno.land?](./what-is-gnolang.md) for the full
picture.

This page walks you from zero to a working local chain and your first
on-chain transaction. Just want the commands? See [Quick Start](./quickstart.md).

:::tip
Try the **[Playground](https://play.gno.land)** to write Gno in your browser.
:::

## Install

> The toolchain has three binaries:

Reviewer: Verified install.md:3 says 5 binaries (gnokey, gno, gnodev, gnobro, gnoweb). Both claims true (one-liner installs 5 but tutorial uses only 3) but the reader has no way to tell.

```suggestion
This tutorial uses three of the toolchain binaries:
```

| Binary    | What it is                                              |
|-----------|---------------------------------------------------------|
| `gno`     | the Gno language toolchain (format, test, run, mod init)|
| `gnokey`  | key management for interacting with the network         |
| `gnodev`  | local development environment with hot reload + web UI  |

```sh
curl -fsSL https://raw.githubusercontent.com/gnolang/gno/master/misc/install.sh | sh
```

Binaries land in `$HOME/.gno/bin`. The script is bash-only; Windows users should use

> WSL or see [Other methods](./install.md) for source builds and Docker.

Reviewer: ```suggestion
WSL or see [Installation page](./install.md) for source builds, Docker, version pinning (--version <tag>), or --full (validator node).
```

Verify the toolchain is on your `$PATH`:

```sh

> gno version && gnokey version && gnodev version

Reviewer: NIT: gnodev version works because this PR ships its own contribs/gnodev/version.go (main.go:67 adds newVersionCmd). Confirmed gnodev version doesn't exist on master. But the same commits also live in OPEN #5563 
Please: coordinate merge order or rebase one onto the other to avoid double-apply conflicts.

Author: This shouldn't cause any conflict after merging #5563 on master

Edit: I removed the dependency for simpler merging

```

:::tip
`gno: command not found`? Add `$HOME/.gno/bin` to your `$PATH`:
`export PATH="$HOME/.gno/bin:$PATH"`.

> :::

Reviewer: - install.md:116 uses export PATH="$PATH:$HOME/.gno/bin" (append)
 - this file uses export PATH="$HOME/.gno/bin:$PATH" (prepend)

Reviewer: Both work: keeping one source of truth is the point of having install.md.

Author: Fix: https://github.com/gnolang/gno/pull/5592/commits/160cf3bf8cd392271b5bd159bff648a4ac5f6e52

> 

Reviewer: Might  be missing: editor setup pointer. UX-2 (#5460) and UX-4 (#5462) both list this?

Author: Fix: https://github.com/gnolang/gno/pull/5553/changes/fd4534c57f7c37b12eacca96d8912f51ebf81f0c
This was pushed on https://github.com/gnolang/gno/pull/5553 (+ added dependency)

## Run locally with gnodev

The fastest way to learn Gno is to run a chain on your own machine.
`gnodev` boots a single-node devnet with hot reload, a built-in
gnoweb UI, and a keybase whose accounts are pre-funded — no faucet,
no genesis file, no peers to configure.

You'll write a small realm, run it through the local toolchain
(`gno fmt`/`lint`/`test`), boot the chain, then sign a real

> transaction against it. Everything in this section runs offline.

Reviewer: Might be missing: "What's different from Go?" callout. Persona A's killer blocker per #5455 is the 1% Go/Gno delta (no goroutines, missing stdlibs, on-chain-only imports). A 2-line callout near the top closes it.

```suggestion
transaction against it. Everything in this section runs offline.

:::tip Coming from Go?
Gno is Go-shaped but not Go. No goroutines, no `os`/`net` stdlibs, on-chain-only imports. See [Go vs Gno compatibility](../resources/go-gno-compatibility.md) for the full delta before you hit a surprise.
:::
```

Author: Fix: https://github.com/gnolang/gno/pull/5592/commits/9b871d0c2e1a5870d8ad037b1c324b655d40490d

### Declare the module path

```sh
gno mod init gno.land/r/myname/myrealm
```

This writes a `gnomod.toml` in the current directory, declaring the
package's on-chain path. Use `gno.land/r/…` for realms (stateful) or
`gno.land/p/…` for pure packages (stateless — note that pure packages cannot

> import realms; see [import rules](../resources/gno-packages.md#import-rules)).

Reviewer: Import realms; see [import rules](./anatomy-of-a-gno-package.md#import-rules)).

Reviewer: The actual "Import Rules" content lives at anatomy-of-a-gno-package.md:19 (### Import Rules).

  ```suggestion
  import realms; see [import rules](./anatomy-of-a-gno-package.md#import-rules)).
  ```

Author: Applied, but this will be modified in https://github.com/gnolang/gno/pull/5656

### Write Gno code

Then, add `.gno` files next to the freshly created `gnomod.toml`. We'll
build a counter — a realm that stores a number and exposes a function to
increment it:

```gno
package myrealm

import "strconv"

var count int

func Increment(_ realm, n int) int {
	count += n
	return count
}

func Render(path string) string {
	return "Count: " + strconv.Itoa(count)
}
```

- **`var count int`** — top-level variables are automatically persisted
  to chain state after each transaction.
- **`Increment`** — the `_ realm` parameter makes the function

>   ["crossing"](../resources/gno-interrealm.md), which is required for

Reviewer: Quoted link text is markdown-renderer-fragile (some renderers leave the quotes literal).

  **Suggestion:**
  ```suggestion
    [crossing function](../resources/gno-interrealm.md), which is required for
  ```

The `gno` CLI ships the same toolchain you'd expect from Go: a
formatter that rewrites code in canonical style, a linter that
catches common mistakes (unused imports, dead code, misuse of
`cross`/`realm`), and a test runner that executes `_test.gno`
files. Run them from the package directory before every commit:

```sh
gno fmt ./...     # rewrite .gno files in canonical style
gno lint ./...    # static checks for common mistakes
gno test ./...    # run _test.gno files
```

### Run a local chain

Once the code passes tests, boot a devnet from the package directory.
`gnodev` starts a single-node chain, loads the realm at its declared
package path, and serves gnoweb in the foreground:

```sh
gnodev .
```

Open http://localhost:8888 — gnoweb shows your realm under its
package path. Click into it to see the `Render` output ("Count: 0"),
browse exported functions and source code, and view prefunded account
balances. Every key in your local `gnokey` keybase is auto-funded at
startup, so no faucet is needed.

Save a `.gno` file and the chain reloads automatically. Pass multiple
directories to load several packages at once; the bundled `examples/`
are loaded by default including

> [`r/docs`](http://localhost:8888/r/docs), an on-chain guided tour you

Reviewer: The precompiled binary delivered by install.sh (this doc's recommended install) has /gnoroot baked in: a path that doesn't exist on the user's machine. The user runs gnodev . and sees no r/docs, only their own realm. The central promise of this section is broken no? However this is important and needs to be fixed if true.

Author: That's right, thank you. Fix: https://github.com/gnolang/gno/pull/5592/commits/160cf3bf8cd392271b5bd159bff648a4ac5f6e52

> can modify and browse locally.

Reviewer: Lines 143-145 claim "Every key in your local gnokey keybase is auto-funded at startup, so no faucet is needed":  but "Create a key" is at L152. A reader following top-to-bottom runs gnodev . before having any key, sees an empty keybase, no test1 (the dev key isn't auto-imported into the user's keybase), and is confused about who "every key" refers to.

Two options: either swap section order (Create a key → Run a local chain), or rephrase to forward-reference.

Author: Only for reference, I've worked on auto-importing the dev key on this PR: https://github.com/gnolang/gno/pull/5680

Author: Fix: https://github.com/gnolang/gno/pull/5592/commits/160cf3bf8cd392271b5bd159bff648a4ac5f6e52

network dropdown and every `gnokey` command's `-remote` and
`-chainid` flags must match:

| Network    | `-chainid` | `-remote`                                  |
|------------|------------|--------------------------------------------|
| Local      | `dev`      | `http://localhost:26657`                   |
| Staging    | `staging`  | `https://rpc.staging.gno.land:443`         |
| Testnet    | `testN`    | `https://rpc.<testN>.testnets.gno.land:443` |

Replace `<testN>` with the current testnet chainid (e.g. `test13`) — see
[Networks](../resources/gnoland-networks.md) for the full list, the
active testnet and mainnet status.

Examples below use **staging** because it resets on a short cadence —
fine for a throwaway first deploy. For anything you want to keep around,
use the current **testnet** instead; staging wipes regularly and your
realm will disappear with it.

### 1. Get test tokens

Deploys cost [gas](../resources/gas-fees.md), paid in `ugnot`. Get test tokens from the faucet:

Go to **[faucet.gno.land](https://faucet.gno.land)**, paste your `g1…`
address, pick a network, and submit. Tokens arrive in seconds. The
faucet is rate-limited per address; wait out the cooldown if a

> re-request is rejected.

Reviewer: Should we mention that we recommend using local for local development, testnet for testing, and mainnet for production only? The faucet will probably not be able to provide tokens for the last case (will have some conditions see [the faucet front last PR](https://github.com/gnolang/faucet-hub/pull/75)).

Author: Fix: https://github.com/gnolang/gno/pull/5592/changes/9b871d0c2e1a5870d8ad037b1c324b655d40490d..0721d1a56c82bd9990150fe55353f1c8bb8408bd
